moz_security/irc.c

```diff
@@ -0,0 +1,550 @@
+        ic->chanqueue = ic->chanqueue->next;
+        free(qp);
+        ic->chanqueue_size--;
+      }
+      pthread_mutex_unlock(&ic->chanqueue_lock);
+    }
+  }
+
+  return (NULL);
+}
+
+static int irclib_bio_write(BIO *b, char *buf, size_t len)
+{
+  for (;;) {
+    if (BIO_write(b, buf, len) <= 0) {
+      if (!BIO_should_retry(b)) {
+        return (-1);
+      }
+      continue;
+    }
+    break;
+  }
+
+  return (0);
+}
+
+static int irclib_tryconn(struct ircconn *ic)
+{
+  char buf[1024];
+
+  memset(ic->inbuf, 0, sizeof(ic->inbuf));
+  ic->inbuf_len = 0;
+  pthread_mutex_lock(&ic->chanqueue_lock);
+  if (ic->chanqueue != NULL) {
+    for (;;) {
+      struct queueent *qp = ic->chanqueue->next;
+      free(ic->chanqueue);
+      if (qp == NULL) {
+        break;
+      }
+      ic->chanqueue = qp;
+    }
+  }
+  ic->chanqueue = NULL;
+  ic->chanqueue_size = 0;
+  pthread_mutex_unlock(&ic->chanqueue_lock);
+
+  ic->bio = BIO_new_ssl_connect(ctx);
+  BIO_set_nbio(ic->bio, 1);
+  snprintf(buf, sizeof(buf), "%s:%d", ic->irc_hn, ic->irc_port);
+  BIO_set_conn_hostname(ic->bio, buf);
+  BIO_get_ssl(ic->bio, &ic->ssl);
+
+  for (;;) {
+    if (BIO_do_connect(ic->bio) != 1) {
+      if (!BIO_should_retry(ic->bio)) {
+        BIO_free_all(ic->bio);
+        return (-1);
+      }
+    } else {
+      break;
+    }
+  }
+  for (;;) {
+    if (BIO_do_handshake(ic->bio) != 1) {
+      if (!BIO_should_retry(ic->bio)) {
+        BIO_free_all(ic->bio);
+        return (-1);
+      }
+    } else {
+      break;
+    }
+  }
+
+  snprintf(buf, sizeof(buf), "NICK %s\r\nUSER %s @ %s :%s\r\n", ic->irc_nick, ic->irc_nick,
+      ic->irc_nick, ic->irc_nick);
+  if (irclib_bio_write(ic->bio, buf, strlen(buf)) == -1) {
+    return (-1);
+  }
+
+  return (0);
+}
+
+static int irclib_new(lua_State *lua)
+{
+  struct ircconn *ic;
+  const char *nick, *hn, *chan, *chankey = NULL;
+  int nargs, port;
+
+  nargs = lua_gettop(lua);
+  luaL_argcheck(lua, nargs >= 4, 0, "incorrect number of arguments");
+
+  nick = luaL_checkstring(lua, 1);
+  hn = luaL_checkstring(lua, 2);
+  port = luaL_checknumber(lua, 3);
+  chan = luaL_checkstring(lua, 4);
+  if (nargs > 4) {
+    chankey = luaL_checkstring(lua, 5);
+  }
+  ic = (struct ircconn *)lua_newuserdata(lua, sizeof(struct ircconn));
+  memset(ic, 0, sizeof(struct ircconn));
+  luaL_getmetatable(lua, MOZSVC_IRC);
+  lua_setmetatable(lua, -2);
+  ic->irc_nick = strdup(nick);
+  ic->irc_hn = strdup(hn);
+  ic->irc_port = port;
+  ic->irc_chan = strdup(chan);
+  if (chankey != NULL) {
+    ic->irc_chankey = strdup(chankey);
+  }
+  ic->status = STATE_DOWN;
+
+  pthread_mutex_init(&ic->chanqueue_lock, NULL);
+  if (pthread_create(&ic->monitor, NULL, irclib_monitor, ic) != 0) {
+    return luaL_error(lua, "pthread_create failed");
+  }
+
+  return (1);
+}
+
+static int irclib_writeraw(lua_State *lua)
+{
+  struct ircconn *ic;
+  char buf[1024];
+  const char *s;
+
+  ic = luaL_checkudata(lua, 1, MOZSVC_IRC);
```

Comment:
You can do the assignment/checks first and fail fast before the 1K buffer is allocated